File: VSCode/compartmental_built.py
## PROMPT: Generate code for an SIR compartmental model using the classes SimulateOde, Transition, TransitionType from PyGOM 
## INTERFACE: VS Code

## PROMPT: Add code which creates a plot to view the outbreak dynamics 

"""SIR compartmental model implemented with PyGOM.

This script builds a simple susceptible-infected-recovered model using
PyGOM's SimulateOde, Transition, and TransitionType classes.
"""

from __future__ import annotations

from typing import Tuple

import numpy as np

from pygom import SimulateOde, Transition, TransitionType


def build_sir_model() -> SimulateOde:
    """Create a PyGOM SIR model object."""
    state = ["S", "I", "R"]
    param_list = ["beta", "gamma", "N"]
    transition = [
        Transition(
            origin="S",
            destination="I",
            equation="beta*S*I/N",
            transition_type=TransitionType.T,
        ),
        Transition(
            origin="I",
            destination="R",
            equation="gamma*I",
            transition_type=TransitionType.T,
        ),
    ]
    return SimulateOde(state, param_list, transition=transition)


def simulate_sir(
    days: float = 200.0,
    dt: float = 0.1,
    beta: float = 0.3,
    gamma: float = 0.1,
    population: float = 1000.0,
    initial_infected: float = 1.0,
) -> Tuple[np.ndarray, np.ndarray]:
    """Simulate the SIR model over time and return the time and state arrays."""
    model = build_sir_model()
    params = {"beta": beta, "gamma": gamma, "N": population}
    model.parameters = params

    times = np.linspace(0.0, float(days), int(float(days) / float(dt)) + 1)
    initial_state = [population - initial_infected, initial_infected, 0.0]
    model.initial_values = (initial_state, times[0])

    solution = np.asarray(model.integrate(times[1:]))
    if solution.ndim == 1:
        solution = solution.reshape(1, -1)

    if solution.shape[0] == len(times):
        solution = solution[1:]
    elif solution.shape[0] != len(times) - 1:
        if solution.shape[0] > len(times) - 1:
            solution = solution[: len(times) - 1]
        else:
            pad = np.repeat(solution[-1:][None, :], len(times) - 1 - solution.shape[0], axis=0)
            solution = np.vstack([solution, pad])

    states = np.vstack([np.array(initial_state, dtype=float)[None, :], solution])

    return times, states

def plot_results(times: np.ndarray, states: np.ndarray) -> None:
    """Plot the SIR outbreak dynamics for the susceptible, infected, and recovered compartments."""
    try:
        import matplotlib.pyplot as plt
    except ImportError:
        print("matplotlib is not installed; skipping plot output.")
        return

    labels = ["S", "I", "R"]
    colors = ["tab:blue", "tab:red", "tab:green"]

    plt.figure(figsize=(8, 4.5))
    for idx, label in enumerate(labels):
        plt.plot(times, states[:, idx], label=label, color=colors[idx], linewidth=1.8)

    plt.xlabel("Time (days)")
    plt.ylabel("Population")
    plt.title("SIR outbreak dynamics")
    plt.grid(alpha=0.3)
    plt.legend(loc="best")
    plt.tight_layout()
    plt.savefig("sir_outbreak_dynamics.png", dpi=200, bbox_inches="tight")
    plt.show()


def main() -> None:
    times, states = simulate_sir()
    labels = ["S", "I", "R"]
    final_state = dict(zip(labels, states[-1]))

    print("SIR model simulation complete.")
    print("Final state:")
    for name, value in final_state.items():
        print(f"  {name}: {value:.2f}")

    infected = states[:, 1]
    peak_day = times[np.argmax(infected)]
    peak_infected = infected.max()
    print(f"Peak infected: {peak_infected:.2f} on day {peak_day:.1f}")

    plot_results(times, states)


if __name__ == "__main__":
    main()

File: VSCode/compartmental_built.R
## PROMPT: Generate code for an SIR compartmental model using odin from R 
## INTERFACE: VS Code

# (PROMPT: Adjust the plot code to accommodate S and R)
# SIR compartmental model implemented with odin in R
# This script simulates a simple susceptible-infectious-recovered epidemic.

if (!requireNamespace("odin", quietly = TRUE)) {
  install.packages("odin")
}

library(odin)

sir_model <- odin::odin({
  initial(S) <- N - I0 - R0
  initial(I) <- I0
  initial(R) <- R0

  beta <- user(0.3)
  gamma <- user(0.1)

  N <- user(1000)
  I0 <- user(1)
  R0 <- user(0)

  lambda <- beta * I / N

  deriv(S) <- -lambda * S
  deriv(I) <- lambda * S - gamma * I
  deriv(R) <- gamma * I

  output(incidence) <- lambda * S
})

model <- sir_model$new(
  N = 1000,
  I0 = 1,
  R0 = 0,
  beta = 0.3,
  gamma = 0.1
)

times <- seq(0, 200, by = 0.1)
solution <- as.data.frame(model$run(times))

matplot(
  solution$t,
  solution[, c("S", "I", "R")],
  type = "l",
  lty = 1,
  lwd = 2,
  col = c("blue", "red", "green"),
  xlab = "Time",
  ylab = "Population",
  main = "SIR model with odin"
)
legend(
  "right",
  legend = c("Susceptible", "Infectious", "Recovered"),
  col = c("blue", "red", "green"),
  lty = 1,
  lwd = 2
)

cat("\nFinal state after 200 time units:\n")
print(tail(solution, 1))

## PROMPT: Add code for a new model that extends the SIR model to account for the key aspects of an Ebola outbreak 

# Ebola-style extension of the SIR model
# Adds exposed, recovered, and deceased compartments to reflect
# latent infection, recovery, and mortality observed in Ebola outbreaks.
ebola_model <- odin::odin({
  initial(S) <- N - E0 - I0 - R0 - D0
  initial(E) <- E0
  initial(I) <- I0
  initial(R) <- R0
  initial(D) <- D0

  beta <- user(0.55)
  sigma <- user(0.35)
  gamma <- user(0.15)
  mu <- user(0.08)

  N <- user(1000000)
  E0 <- user(0)
  I0 <- user(1)
  R0 <- user(0)
  D0 <- user(0)

  lambda <- beta * I / N

  deriv(S) <- -lambda * S
  deriv(E) <- lambda * S - sigma * E
  deriv(I) <- sigma * E - (gamma + mu) * I
  deriv(R) <- gamma * I
  deriv(D) <- mu * I

  output(incidence) <- lambda * S
})

model_ebola <- ebola_model$new(
  N = 1000000,
  E0 = 0,
  I0 = 1,
  R0 = 0,
  D0 = 0,
  beta = 0.55,
  sigma = 0.35,
  gamma = 0.15,
  mu = 0.08
)

times_ebola <- seq(0, 180, by = 1)
solution_ebola <- as.data.frame(model_ebola$run(times_ebola))

matplot(
  solution_ebola$t,
  solution_ebola[, c("S", "E", "I", "R", "D")],
  type = "l",
  lty = 1,
  lwd = 2,
  col = c("blue", "orange", "red", "green", "black"),
  xlab = "Time (days)",
  ylab = "Population",
  main = "Ebola-style extension of the SIR model"
)
legend(
  "right",
  legend = c("Susceptible", "Exposed", "Infectious", "Recovered", "Deceased"),
  col = c("blue", "orange", "red", "green", "black"),
  lty = 1,
  lwd = 2
)

cat("\nFinal state after 180 days:\n")
print(tail(solution_ebola, 1))
